src/coreclr/jit/emitxarch.h

```diff
@@ -160,6 +168,249 @@ code_t AddVexPrefixIfNeededAndNotPresent(instruction ins, code_t code, emitAttr
+        case INS_punpckhdq:
+        case INS_addps:
+        case INS_addss:
+        case INS_mulss:
+        case INS_mulps:
+        case INS_subss:
+        case INS_subps:
+        case INS_minss:
+        case INS_minps:
+        case INS_divss:
+        case INS_divps:
+        case INS_maxss:
+        case INS_maxps:
+        case INS_xorps:
+        case INS_andps:
+        case INS_sqrtss:
+        case INS_sqrtps:
+        case INS_andnps:
+        case INS_orps:
+        case INS_cvtss2sd:
+        case INS_cvtdq2ps:
+        case INS_cvtps2dq:
+        case INS_cvttps2dq:
+        case INS_cvtdq2pd:
+        case INS_paddd:
+        case INS_psubd:
+        case INS_pslld:
+        case INS_psrld:
+        case INS_psrad:
+        case INS_pshufd:
+        case INS_packssdw:
+        case INS_insertps:
+        case INS_pmulld:
+        case INS_pabsd:
+        case INS_pminsd:
+        case INS_pminud:
+        case INS_pmaxud:
+        case INS_pmovsxdq:
+        case INS_pmovzxdq:
+        case INS_packusdw:
+        case INS_movntdqa:
+        case INS_movsldup:
+        case INS_movshdup:
+        case INS_pinsrd:
+        case INS_pextrd:
+        case INS_vbroadcastss:
+        case INS_vbroadcastsd:
+        case INS_vpbroadcastb:
+        case INS_vpbroadcastw:
+        case INS_vpbroadcastd:
+        case INS_vpsravd:
+        case INS_vpsllvd:
+        case INS_vpermilps:
+        case INS_vpermd:
+        case INS_vpermps:
+        case INS_vpgatherdd:
+        case INS_vpgatherqd:
+        case INS_vgatherdps:
+        case INS_vgatherqps:
+        case INS_vfmadd132ps:
+        case INS_vfmadd213ps:
+        case INS_vfmadd231ps:
+        case INS_vfmadd132ss:
+        case INS_vfmadd213ss:
+        case INS_vfmadd231ss:
+        case INS_vfmaddsub132ps:
+        case INS_vfmaddsub213ps:
+        case INS_vfmaddsub231ps:
+        case INS_vfmsubadd132ps:
+        case INS_vfmsubadd213ps:
+        case INS_vfmsubadd231ps:
+        case INS_vfmsub132ps:
+        case INS_vfmsub213ps:
+        case INS_vfmsub231ps:
+        case INS_vfmsub132ss:
+        case INS_vfmsub213ss:
+        case INS_vfmsub231ss:
+        case INS_vfnmadd132ps:
+        case INS_vfnmadd213ps:
+        case INS_vfnmadd231ps:
+        case INS_vfnmadd132ss:
+        case INS_vfnmadd213ss:
+        case INS_vfnmadd231ss:
+        case INS_vfnmsub132ps:
+        case INS_vfnmsub213ps:
+        case INS_vfnmsub231ps:
+        case INS_vfnmsub132ss:
+        case INS_vfnmsub213ss:
+        case INS_vfnmsub231ss:
+        case INS_vpdpbusd:
+        case INS_vpdpwssd:
+        case INS_vpdpbusds:
+        case INS_vpdpwssds:
+        case INS_vpermilpsvar:
+        {
+            return false; // W0
+        }
+        default:
+        {
+            return false; // WIG
+        }
+    }
+}
+
+// TODO-XArch-AVX512: Currently kmask is not supported.
+// Refactor once kmask is added.
+bool HasKMaskRegisterDest(instruction ins) const
+{
```

Comment:
This should probably have enough info to assert it is only true when emitting the EVEX encoding.

Reply:
Eventually yes. But since right now, we do not have kmask support, we ae using this to make the decision of IF we can EVEX encode these. https://github.com/DeepakRajendrakumaran/runtime/blob/avx512-RRR/src/coreclr/jit/emitxarch.cpp#L720-L733. So, I'm not sure what check we can add here right now. Maybe `UseEVEXEncoding() == true` ?

Reply:
Right, I think we want to ensure this path is only called when `EVEX` is enabled and ideally only when the instruction is explicitly marked as wanting the KMASK variant.

When we aren't EVEX -or- when we are EVEX but we're using one of the overloads that returns a SIMD mask instead, then we want to ensure this is "false".

Reply:
Just so that we are on the same page, you mean  `assert(UseEVEXEncoding() == true)` within `HasKMaskRegisterDest(ins)` ?

The current order of operations make it a little weird. So, how we determine if `TakesEvexPrefix() == true`

Step 1 : If the JitStressEVEXEncoding flag is false, return false
Step 2 : If `HasKMaskRegisterDest(ins)` is true, return false
Step 3 : If both above conditions are used to rule out instructions for EVEX encoding. If we make it here, use `IsAVX512Instruction(ins)`

Within `IsAVX512Instruction(ins)`, we check if `UseEVEXEncoding()` is true among other conditions to determine whether to Evex encode. The reason I lay this out is, because of this order of operation, within `HasKMaskRegisterDest(ins)`, it won't always satisfy the condition `UseEVEXEncoding() == true` The reason i have it this way is basically following pre-existing VEX code

```
bool emitter::IsAVXInstruction(instruction ins) const
{
    return UseVEXEncoding() && IsSSEOrAVXInstruction(ins);
}
```

A simple solution would be to move the `IsAVX512Instruction(ins)` check to the top of the function This is ensure that we can add the assert in `HasKMaskRegisterDest(ins)`  . **I'll make this change**

Reply:
The concern I'm raising here is that we have some instructions which **require** `kmask` destinations and some which **can use** `kmask` destinations.

For the instructions like `pcmpeqb` we don't want to hit this path when emitting `Vector128<byte> Sse2.CompareEqual(Vector128<byte>, Vector128<byte>)` since that, even when `EVEX` can be used, is exclusively returning an `XMM` register.

Instead, we'd only want to use it for something like `VectorMask128<byte> Avx512F.VL.CompareEqual(Vector128<byte>, Vector128<byte>)`

Reply:
Ah. I see what you're saying.
 
Just for clarification `HasKMaskRegisterDest(ins)` is not meant to sort through those differences. It's a temporary check which needs to be removed once we add k-mask support. It just identifies intrinsics for which EVEX encoding exists but **REQUIRES** k mask if doing EVEX encoding. So, even when we are forcing EVEX encoding with `JitStressEVEXEncoding` flag, this just tells us to skip these instructions **for now till we support k mask**.

`pcmpeqb` is  a good example. My understanding is, if using EVEX encoding, the result **REQUIRES** k mask since the result is written to kmask(https://www.felixcloutier.com/x86/pcmpeqb:pcmpeqw:pcmpeqd). Meaning we cannot EVEX encode `pcmpeqb`

PCMPEQB **k1 {k2}**, xmm2, xmm3 /m128.


`Instead, we'd only want to use it for something like VectorMask128<byte> Avx512F.VL.CompareEqual(Vector128<byte>, Vector128<byte>)`

The actual logic for differentiating instructions which can use EVEX but does not require EVEX should be in here but added currently. It's noted in comments but not extrapolated

https://github.com/DeepakRajendrakumaran/runtime/blob/avx512-RRR/src/coreclr/jit/emitxarch.cpp#L709-L733

p.s. Hopefully, I understood your concern correctly and I'm addressing the question. Let me know if I'm answering the wrong questions :)

Reply:
Thanks, this makes sense.